### PR TITLE
feat: Add authentication for ACP

### LINKED ISSUE #2017: ACP - Full Identity Authentication
ACP - Full Identity Authentication

Initially we only have simple identity

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2017

## Description

This PR adds ACP identity authentication via HTTP.

Notable changes:
- `acp/identity` has been replaced with the `acp.Identity` struct
  - `identity.PrivateKey` is the private key of the identity
  - `identity.PublicKey` is the public key of the identity
  - `identity.Address` is the bech32 formatted address for the identity
  - keys are all `secp256k1`
- `http` can authenticate requests using a jwt bearer token
  - ~a random `audience` value is generated on every http server startup~
  - audience must be set to the defradb host name
  - ~api route `/audience` returns the random audience value~
  - `http.Client` will create a signed token if an `acp.PrivateKeyIdentity` is set
  - jwt token subject is the identity public key
- `cli` `--identity` flag is now a hex encoded private key

Todo:
  - [x] ensure acp docs are updated

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

`make test`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: acp/identity.go
Comment (AndrewSisley) [inline acp/identity.go:27]: suggestion: The callers of this function (with a lazy ctrl+f in the browser) are in relatively generalized parts of the codebase where maintainers do not need to understand or know what a `bech32 address` and why that is significant.

At the moment this comment/documentation does not describe if or why it being `bech32` is significant, and the calling sites appear to be ignorant of that too.

If its `bech32`ness is important, please describe why.  If it is not I suggest either removing it from the comment, or reducing its importance by moving it onto a secondary line (e.g. below).
```
// Address returns the address of the identity.  // the address where (maybe also note if this is in sourcehub/keyring/whereever)
//
// note on it being in bech32 format
```

File: acp/identity.go
Comment (AndrewSisley) [inline acp/identity.go:26]: question: `secp256k1.PublicKey` is a very specific type to be returning from an otherwise quite generic looking interface, is there another slightly more abstract type that we can return instead to save coupling all consumers of this function to a very specific key type?

File: http/handler.go
Comment (AndrewSisley) [inline http/handler.go:97]: todo: Please document why it is safe to ignore this error, if it is safe.

File: tests/integration/acp/add_policy/with_invalid_creator_arg_test.go
Comment (AndrewSisley) [inline tests/integration/acp/add_policy/with_invalid_creator_arg_test.go:55]: todo: You appear to have changed the test so that it no longer tests what the name and description say it does.

The linked ticket https://github.com/sourcenetwork/defradb/issues/2357 suggests that this is because of a sourcehub bug, which presumably is not fixed by this PR.

Is it now impossible to provide an invalid identity to defra? (this seems unlikely to me).  If it is still possible to provide an invalid identity, please preserve this testing what it was.   Otherwise it might be best to delete it and close the ticket.

Please reply to this message with what you have done when you have done it, so I don't forget or lose track of it :)

File: http/auth_test.go
Comment (AndrewSisley) [inline http/auth_test.go:26]: question: How is authentication tested?  Is it just this file, or is it automagically covered by the http/cli integration tests?  Is authentication now always required when ACP is enabled?  Is there now no easy way for users of the embedded Go client to provide an ACP identity (is this untested?)?

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Looks good, but I have a handful of questions/todos for you at the moment :)

File: acp/identity.go
Comment (AndrewSisley) [inline acp/identity.go:28]: question: Why are you returning a `bech32` string instead of a `"github.com/cosmos/cosmos-sdk/crypto/types"`.`Address`?

File: tests/integration/acp/add_policy/with_empty_args_test.go
Comment (shahzadlone) [inline tests/integration/acp/add_policy/with_empty_args_test.go:45]: todo: I don't like these changes, the tests should explicitly show the identity being empty.

The previous tests should stay as much as possible in my opinion. I think what you want to test in this PR and what the previous tests test are 2 fundamentally separate things.

The previous tests assert that the RAW identity (without being wrapped in auth token) work. Which should continue behaving like they did.

You should introduce a new Variable called `IdentityWithAuth` within the respective `testUtils.XYZ` util functions where you want to test the authentication, then copy the tests you think are relavent from `tests/integration/acp` and add them to a dir called `test/integration/acp/auth`. This way we don't mix the testing of authentication and the acp logic.

So al/mostl the previous tests should be reverted and unchanged in this PR and new one introduced with using `IdentityWithAuth` instead of `Identity` (you this way also don't have to re-test  all acp cases, but only the authentication functionality.

**Review by shahzadlone — CHANGES_REQUESTED** [review]

File: acp/identity/identity.go
Comment (shahzadlone) [inline acp/identity/identity.go:1]: issue/conflict: You have removed the identity package, while my branch has expanded it (with some mostly similar functionality to your identity stuff). We should definitely coordinate on the structure.

I was thinking to actually more the `indentity` package perhaps out of the `acp` package into it's own thing. Can discuss that.

File: acp/identity.go
Comment (shahzadlone) [inline acp/identity.go:54]: todo: I don't like this name it's a bit confusing as to what different type of identities there are now? I think all you wanted was a struct for private key?

File: tests/integration/acp/add_policy/with_empty_args_test.go
Comment (AndrewSisley) [reply tests/integration/acp/add_policy/with_empty_args_test.go:45]: > The previous tests assert that the RAW identity (without being wrapped in auth token) work. Which should continue behaving like they did.

This is related to one of my questions I think.  We might want to chat as a team as to whether we want to support unauthenticated identities in the future.  The jwt token is also required for SourceHub as far as I can tell, and the usefulness of local-ACP is debated atm.

File: tests/integration/acp/add_policy/with_invalid_creator_arg_test.go
Comment (AndrewSisley) [reply tests/integration/acp/add_policy/with_invalid_creator_arg_test.go:55]: > The linked ticket https://github.com/sourcenetwork/defradb/issues/2357 suggests that this is because of a sourcehub bug, which presumably is not fixed by this PR.

note: This is actually fixed in the `feat/sdk-pkg` branch of SourceHub, and maybe the latest `dev` branch

File: acp/identity.go
Comment (fredcarle) [reply acp/identity.go:28]: I think it's because `acpSystem.RegisterDocObject` takes a string and that method is used for that.

File: acp/identity.go
Comment (shahzadlone) [inline acp/identity.go:24]: conflict: My branch introduces similar functionality but instead of using `Identity` as an interface and then two different structs implementing it, there is just one `Identity` struct that has three components:

``` go
// Identity holds the bits that make up an actor's unique identity.
type Identity struct {
	// An actor's private key. Uses secp256k1 algorithm.
	PrivateJWK string

	// An actor's corresponding public key address.
	PublicKey string

	// An actor's DID. Generated from the public key address.
	DID string
}
```

Note: `did` is what it will be after my stuff merges not a sourcehub address.

IMO we don't need to over abstract the identity with an interface, just a struct like above would suffice where what would have as `PrivateKeyIdentity` would just be the `Identity` struct above with all non-empty fields, and `PublicKeyIdentity` would the `Identity` struct above with just `PrivateJWK` as empty. One liner documenting this on the struct would make it clear between these differences.

todo: unabstract interface identity implementation

File: crypto/crypto.go
Comment (shahzadlone) [inline crypto/crypto.go:17]: discussion:

I think I was using `	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"`

Any differences you know between them?

File: crypto/crypto.go
Comment (shahzadlone) [inline crypto/crypto.go:21]: question: You only need this for testing right? not for any production code?

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Did a first pass quickly, left some todos, will review again tonight.

Can you please also update the acp readme with authentication cli examples now that we will always have acp with authentication. It will help better understand for reviewers as well, how acp with auth works.

File: tests/integration/acp/add_policy/with_empty_args_test.go
Comment (shahzadlone) [reply tests/integration/acp/add_policy/with_empty_args_test.go:45]: When I said 

> Which should continue behaving like they did.

I meant the access control part should behave normally as it did, sorry. Not that we should keep non-auth identity functionality.

In the comment above I am suggesting to not mix the "authentication" testing with the "access control" testing, and keep them separate.

As for weather we want unauthenticated identities, I really can't think of a good reason why we would want to keep that if we have auth implemented.

As for:

> the usefulness of local-ACP is debated atm

I don't see how it's debated haha, IMO is really useful for anyone who wants to use defradb like a classic/normal use case (i.e. without a sourcehub-like instance). We can test in standup though, maybe I am missing something.

File: acp/identity.go
Comment (shahzadlone) [reply acp/identity.go:28]: I don't think this matters as address will be replaced with `did` in a follow up work

File: tests/integration/acp/add_policy/with_empty_args_test.go
Comment (AndrewSisley) [reply tests/integration/acp/add_policy/with_empty_args_test.go:45]: > In the comment above I am suggesting to not mix the "authentication" testing with the "access control" testing, and keep them separate.

How would you keep them separate without permitting auth-free identities?

File: tests/integration/acp/add_policy/with_invalid_creator_arg_test.go
Comment (shahzadlone) [reply tests/integration/acp/add_policy/with_invalid_creator_arg_test.go:55]: When we bump to latest sourcehub commit on `dev` it I believe fixes this panic.

File: acp/identity.go
Comment (nasdf) [reply acp/identity.go:50]: Since we are going between well known formats this should never panic

Comment (AndrewSisley) [reply acp/identity.go:50]: Can you please add a comment explaining that :)

File: http/auth_test.go
Comment (nasdf) [reply http/auth_test.go:26]: > How is authentication tested?

It's tested by the integration tests as well. Any place we are using an identity in a test will use authentication when using the http / cli clients.

> Is there now no easy way for users of the embedded Go client to provide an ACP identity (is this untested?)?

This doesn't change anything with how we use identity in the Go client.

```go
identity := acp.IdentityFromPublicKey(...)
ctx := db.SetContextIdentity(ctx, identity)
```

File: tests/integration/acp/add_policy/with_empty_args_test.go
Comment (nasdf) [reply tests/integration/acp/add_policy/with_empty_args_test.go:45]: > The previous tests assert that the RAW identity (without being wrapped in auth token) work. Which should continue behaving like they did.

The Go client doesn't create auth tokens. Only the HTTP / CLI clients use auth tokens.

The change here was made because the new identity requires a valid public or private key and an empty string is not a valid identity.

File: acp/identity.go
Comment (AndrewSisley) [inline acp/identity.go:34]: question: I can't spot where this is used, is it dead code?

File: crypto/crypto.go
Comment (nasdf) [reply crypto/crypto.go:21]: Yes, I'm planning on adding a way to generate and store them in the keyring as well.

File: acp/identity.go
Comment (nasdf) [reply acp/identity.go:34]: It's a client facing utility, but will be removed when I merge in Shahzad's identity changes.

Comment (AndrewSisley) [reply acp/identity.go:34]: What will replace it?  Is there any value to it living for just the (short?) interval between merging this PR and Shahzad's changes?

I'm asking partly because SourceHub ACP needs stuff to be signed (client-side), which requires the private key.

Comment (nasdf) [reply acp/identity.go:34]: > What will replace it?

Something like this

```go
type Identity struct {
	// PublicKey is the identity public key.
	PublicKey *secp256k1.PublicKey
	// PrivateKey is the identity private key.
	PrivateKey *secp256k1.PrivateKey
	// Address is the identity address.
	Address string
}
```

> Is there any value to it living for just the (short?) interval between merging this PR and Shahzad's changes?

It's a quick change and Shahzad will most likely want this done before approving.

File: acp/identity.go
Comment (nasdf) [reply acp/identity.go]: I've flattened like the identity interface to struct like you suggested. Let me know which package you would like for it to exist in.

File: acp/identity.go
Comment (nasdf) [reply acp/identity.go]: The keys will be changed to a JWK shortly after this PR. It's probably not worth the effort to change them in the mean time.

**Review by shahzadlone — COMMENTED** [review]
Some more questions.

Question: Was the ability to send requests without an identity (optionality of an identity being set or specified) removed? IMO we need to maintain optional identity.

Question: If Identity is still optional, when a verification fails, do you error out or assume that no identity was provided and still execute request? IMO should be an error even if optional identity is stripped.

File: internal/db/db.go
Comment (nasdf) [reply internal/db/db.go]: This should be the same behavior as before. If the identity is empty the creator address will be an empty string.

File: tests/clients/cli/wrapper_cli.go
Comment (nasdf) [reply tests/clients/cli/wrapper_cli.go:61]: Since identity is a struct it will always have a default value. In this case we only care about the private key being non nil.

**Comment by nasdf** [comment]
> Question: Was the ability to send requests without an identity (optionality of an identity being set or specified) removed? IMO we need to maintain optional identity.

No

> Question: If Identity is still optional, when a verification fails, do you error out or assume that no identity was provided and still execute request? IMO should be an error even if optional identity is stripped.

If verification fails the request returns a `403` status code / forbidden error.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Re-requesting changes, you appear to have forgotten about/forgotten to push a few of my previous todos.

File: tests/integration/acp/add_policy/with_invalid_creator_arg_test.go
Comment (nasdf) [reply tests/integration/acp/add_policy/with_invalid_creator_arg_test.go]: I've restored the tests but limited them to go client only.

Comment (AndrewSisley) [reply tests/integration/acp/add_policy/with_invalid_creator_arg_test.go]: Thanks Keenan, and thanks for the code-comments explaining why :)

File: acp/identity/identity.go
Comment (AndrewSisley) [inline acp/identity/identity.go:54]: thought:  This looks like it might actually panic if `"cosmos"` is no longer the prefix (it is actually being changed at the moment, so that's not without precedent).

todo: There is a const for this that you can use instead of a magic string, please use `"github.com/sourcenetwork/sourcehub/app"`.`AccountAddressPrefix`

File: http/handler.go
Comment (nasdf) [reply http/handler.go]: I've updated both response writer utils to log any errors.

File: acp/identity/identity.go
Comment (nasdf) [reply acp/identity/identity.go:54]: Using that causes a panic `panic: invalid Bech32 prefix; expected cosmos, got source`

Comment (AndrewSisley) [reply acp/identity/identity.go:54]: Ah okay, using the const might have to wait on Shahzad's PR then, and I've been chatting with Bruno today - he might change the source hub sdk so that Defra controls the prefix (in which case we can either own a const, or have a config option for this).

**Review by AndrewSisley — APPROVED** [review]
LGTM, thanks Keenan! I would prefer that you wait on Shahzad's approval before merge though :)

File: acp/identity/identity.go
Comment (nasdf) [reply acp/identity/identity.go:16]: Discussed in call: `github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1` will be removed when we switch to DIDs.

File: cli/acp_policy_add.go
Comment (nasdf) [reply cli/acp_policy_add.go]: I've fixed the linter error issues.

File: acp/identity/identity.go
Comment (nasdf) [reply acp/identity/identity.go]: I've updated the documentation here as suggested

File: acp/identity/identity.go
Comment (nasdf) [reply acp/identity/identity.go]: I've updated the documentation here as suggested

File: acp/identity/identity.go
Comment (nasdf) [reply acp/identity/identity.go]: I've updated the documentation here as suggested

**Comment by jsimnz** [comment]
reviewing :) 👀

File: tests/clients/cli/wrapper_cli.go
Comment (jsimnz) [reply tests/clients/cli/wrapper_cli.go:61]: the point of the `Option` is to express the zero/empty value explicitly, rather than implicitly with Go zero values (like a default struct).

If we're going to wrap the identity in the `Option` type, we should at least use it correctly.

**Review by jsimnz — APPROVED** [review]
Obviously there is on going work with this part of the codebase, approving now so I don't further block the merging. Only one todo, and some thoughts/questions

File: acp/identity/identity.go
Comment (shahzadlone) [reply acp/identity/identity.go:16]: > Discussed in call: `github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1` will be removed when we switch to DIDs.

But we still need it for the one case of the hardcoded source address so I think it will remain :/

File: acp/README.md
Comment (nasdf) [reply acp/README.md:179]: I'm planning on adding keyring support for identity keys in a follow up PR. I think Shahzad is working on a command that will generate an identity with a private key as well.

File: http/handler.go
Comment (nasdf) [reply http/handler.go:80]: The problem I ran into is that the peer ID can't easily be retrieved when p2p is disabled.

File: acp/README.md
Comment (shahzadlone) [reply acp/README.md:179]: Thinking again the `:` is not nice. I am wondering if the `util` command should generate a `Hex` version of the private key as well. I do know that it's temporary so perhaps until key ring is implemented the `Hex` version of private key can be in the util command.

Comment (shahzadlone) [reply acp/README.md:179]: technically can also use this command, to get the trimmed hex without `:` private key:

```
openssl ecparam -name secp256k1 -genkey | openssl ec -text -noout | head -n5 | tail -n3 | tr -d '\n:\ '
```

File: tests/integration/acp/fixture.go
Comment (shahzadlone) [inline tests/integration/acp/fixture.go:27]: question: Why does this need to be in an `init` function?

todo: I would much prefer a util function some what like this:

```go
func toIdentity(privateHex string) immutable.Option[acpIdentity.Identity] {
	privateKeyBytes, err := hex.DecodeString(privateHex)
	if err != nil {
		panic(err)
	}

	return acpIdentity.FromPrivateKey(
		secp256k1.PrivKeyFromBytes(
			privateKeyBytes,
		),
	)
}

var (
	Actor1Identity = toIdentity("028d53f37a19afb9a0dbc5b4be30c65731479ee8cfa0c9bc8f8bf198cc3c075f")
	Actor2Identity = toIdentity("4d092126012ebaf56161716018a71630d99443d9d5217e9d8502bb5c5456f2c5")
)

```

suggestion: Something similar can replace the `loadIdentity` function you have in the other test as well.

File: tests/integration/acp/index/fixture.go
Comment (shahzadlone) [inline tests/integration/acp/index/fixture.go:42]: praise: Thanks for maintaining the documentation accuracy here, would also be worth fixing the user facing examples in the example directory and the acp readme as well (I think I mentioned in other comments).

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Thanks Keenen for doing the changes. I have some more comments that I have posted, please ping me once ready for re-review, and will give another final review :) Looks really nice now and I am super excited to have this in.

Sorry I forgot to post the action item points from our zoom call yesterday, I know you have done some of those items, but just leaving them here for reference to knock them all down pre-merge.

- Explicit empty identity in the tests that it was removed from
- Use identity option before checking private key, or re-think the option data struct now that identity must have private before verification and will only have public key after verfication.
- Documentation requests:
    - Document how to generate hex from private key for CLI
    - Document why we use HEX for CLI
    - Document HEX type for decoding and encoding
    - Document HTTP auth-flow 
    - Document command using `/audiance` endpoint/domain for HTTP
    - Document how to get the signed token with HTTP
    - Document where to put the signed token with HTTP (i.e. `Bearer` often times people confuse this with GQL Variable haha)

File: net/peer_test.go
Comment (shahzadlone) [reply net/peer_test.go:1093]: or a helper like `toIdentity` I suggested in: https://github.com/sourcenetwork/defradb/pull/2649#discussion_r1621484244

File: cli/collection_delete.go
Comment (nasdf) [reply cli/collection_delete.go:32]: It works for me

File: http/handler.go
Comment (nasdf) [reply http/handler.go]: Audience is now the api host name. Will update to include peer ID in a follow up PR.

**Review by shahzadlone — APPROVED** [review]
Awesome work @nasdf appreciate you doing all the changes :)

One last nitpick, before merge if you can change the title of the PR to have the first word as an action verb instead of `ACP` it would be nice.

For example, `feat: Add Authentication for ACP`

**Pull request merged**
